code/datums/status_effects/debuffs/fire_stacks.dm

```diff
@@ -0,0 +1,379 @@
+		leg_clothes = victim.w_uniform
+	if(leg_clothes)
+		burning_items |= leg_clothes
+
+	for(var/obj/item/burning in burning_items)
+		burning.fire_act((stacks * 25 * delta_time)) //damage taken is reduced to 2% of this value by fire_act()
+
+/**
+ * Handles mob ignition, should be the only way to set on_fire to TRUE
+ *
+ * Arguments:
+ * - silent: When set to TRUE, no message is displayed
+ *
+ */
+
+/datum/status_effect/fire_handler/fire_stacks/proc/ignite(silent = FALSE)
+	if(HAS_TRAIT(owner, TRAIT_NOFIRE))
+		return FALSE
+
+	on_fire = TRUE
+	if(!silent)
+		owner.visible_message(span_warning("[owner] catches fire!"), span_userdanger("You're set on fire!"))
+
+	if(firelight_type)
+		firelight_ref = WEAKREF(new firelight_type(owner))
+
+	SEND_SIGNAL(owner, COMSIG_LIVING_IGNITED, owner)
+	cache_stacks()
+	update_overlay()
+
+/**
+ * Handles mob extinguishing, should be the only way to set on_fire to FALSE
+ */
+
+/datum/status_effect/fire_handler/fire_stacks/proc/extinguish()
+	if(firelight_ref)
+		var/atom/firelight = firelight_ref.resolve()
+		if(firelight && !QDELETED(firelight))
+			qdel(firelight)
+
+	on_fire = FALSE
+	SEND_SIGNAL(owner, COMSIG_CLEAR_MOOD_EVENT, "on_fire")
+	SEND_SIGNAL(owner, COMSIG_LIVING_EXTINGUISHED, owner)
+	cache_stacks()
+	update_overlay()
+
+/datum/status_effect/fire_handler/fire_stacks/on_remove()
+	if(on_fire)
+		extinguish()
+	set_stacks(0)
+	update_overlay()
+
+/datum/status_effect/fire_handler/fire_stacks/update_overlay()
+	if(iscyborg(owner))
+		var/fire_icon = "generic_burning[get_special_icon()]"
+		var/mutable_appearance/fire_overlay = mutable_appearance('icons/mob/onfire.dmi', fire_icon)
+		if(stacks && on_fire)
+			if(last_icon_state != fire_icon)
+				owner.add_overlay(fire_overlay)
+				last_icon_state = fire_icon
+		else
+			if(last_icon_state)
+				owner.cut_overlay(fire_overlay)
+				last_icon_state = null
+		return TRUE
+
+	if(!iscarbon(owner))
+		return FALSE
+
+	var/mob/living/carbon/victim = owner
+	var/fire_icon = "generic_burning[get_special_icon()]"
+	if(ishuman(victim) && stacks > HUMAN_FIRE_STACK_ICON_NUM)
+		var/mob/living/carbon/human/human_victim = victim
+		if(human_victim.dna && human_victim.dna.species)
+			fire_icon = "[human_victim.dna.species.fire_overlay][get_special_icon()]"
+		else
+			fire_icon = "human_burning[get_special_icon()]"
+
+	if((stacks > 0 && on_fire) || HAS_TRAIT(victim, TRAIT_PERMANENTLY_ONFIRE))
+		if(fire_icon == last_icon_state)
+			return TRUE
+
+		victim.remove_overlay(FIRE_LAYER)
+		var/mutable_appearance/new_fire_overlay = mutable_appearance('icons/mob/onfire.dmi', fire_icon, -FIRE_LAYER)
+		new_fire_overlay.appearance_flags = RESET_COLOR
+		victim.overlays_standing[FIRE_LAYER] = new_fire_overlay
+		victim.apply_overlay(FIRE_LAYER)
+		last_icon_state = fire_icon
+
+	else if(last_icon_state)
+		victim.remove_overlay(FIRE_LAYER)
+		victim.apply_overlay(FIRE_LAYER)
+		last_icon_state = null
+
+	return TRUE
```

Comment:
I still have a problem with this
If you're going to be sticking istypes in this proc, it should be handled instead by proc overrides
the same is true of a lot of this code

that doesn't mean you need to drop the modularity, I just want the override logic handled by overrides, and not interwoven istypes